doc/sphinx/chpl2rst.py

```diff
@@ -0,0 +1,219 @@
+    output.append('='*len(basename))
+
+    with open(chapelfile, 'r') as handle:
+        commentdepth = 0
+        state = ''
+        indentation = -1
+
+        # Each line is rst or code-block
+        for line in [l.strip('\n') for l in handle.readlines()]:
+
+            # Skip empty lines
+            if len(line.strip()) == 0:
+                output.append('')
+                continue
+
+            # Comment canaries
+            # TODO -- support escaped characters (to make mppf proud)
+            commentstarts = line.count('/*')
+            commentends = line.count('*/')
+            commentdepth += commentstarts - commentends
+
+            # State tracking
+            laststate = state
+            state = ''
+
+            # Identification of line
+            if commentdepth > 0 or commentends > 0:
+                state = 'blockcomment'
+            elif line.startswith('//'):
+                state = 'linecomment'
+            elif 'code' in laststate:
+                state = 'code'
+            else:
+                state = 'codeblock'
+
+            if 'comment' in state:
+
+                if 'comment' not in laststate:
+                    output.append('')
+
+                rstline = line
+                if state == 'linecomment':
+                    # Strip white space for line comments
+                    rstline = rstline.replace('//', '  ')
+                    rstline = rstline.strip()
+                else:
+                    # Preserve white space for block comments
+                    if commentstarts:
+                        rstline = rstline.replace('/*', '  ')
+                    if commentends > 0:
+                        rstline = rstline.replace('*/', '  ')
+                    if '.. code-block::' in rstline or len(rstline.strip()) == 0:
+                        rstline = rstline.strip()
+
+                # Strip indentation
+                if indentation == -1:
+                    lineindentation = len(rstline) - len(rstline.lstrip(' '))
+                    if lineindentation > 0:
+                        indentation = lineindentation
+                        rstline = rstline.lstrip(' ')
+                else:
+                    if rstline.startswith(' '*indentation):
+                        rstline = rstline[indentation:]
+                    else:
+                        rstline = rstline.lstrip(' ')
+
+                output.append(rstline)
+            else:
+                # Reset indentation
+                indentation = -1
+
+                # Write code block
+                if state == 'codeblock':
+                    output.append('')
+                    output.append('.. code-block:: chapel')
+                    output.append('')
+                codeline = ''.join(['    ', line])
+                output.append(codeline)
+
+    return '\n'.join(output)
+
+
+def chpl2codeblock(chapelfile):
+    """Convert Chapel file to giant code block rst file"""
```

Comment:
Technically it converts it to a string that can then be written to a file write? Just based on the comment I assumed the function would write to the file as well.